BBPebbleLib/Pebble.cpp

```diff
@@ -38,46 +43,86 @@ void Pebble::onDisconnect()
+
+        // Sanity checks on the message_length
+        if (message_length == 0) {
+            qDebug() << "received empty message";
+            bt_socket.read(header_length); // skip this header
+            continue; // check if there are additional headers.
+        } else if (message_length > 8 * 1024) {
+            // Protocol does not allow messages more than 8K long, seemingly.
+            qDebug() << "received message size too long: " << message_length;
+            bt_socket.readAll(); // drop entire input buffer
             return;
         }
-        in >> this->payloadSize;
-        in >> this->endPoint;
-    }
 
-    if(bt_socket.bytesAvailable() < this->payloadSize){
-        return;
-    }
-    char buffer[this->payloadSize];
-    in.readRawData(buffer, this->payloadSize);
-    QByteArray payload(buffer);
+        // Now wait for the entire message
+        if (bt_socket.bytesAvailable() < header_length + message_length) {
+            qDebug() << "incomplete msg body in read buffer";
+            return; // try again once more data comes in
+        }
 
-    qDebug() << "Received data. Endpoint : " << endPoint << " Payload Size : " << payloadSize << " HEX : " << payload.toHex() << "STR : " << QString(payload);
+        // We can now safely remove the header from the input buffer,
+        // as we know the entire message is in the input buffer.
+        bt_socket.read(header_length);
 
-    emit onDataReadFinished(endPoint, payload);
+        // Now read the rest of the message
+        QByteArray data = bt_socket.read(message_length);
 
-    this->payloadSize = 0;
-    this->endPoint = 0;
+        qDebug() << "<<<< received data. Endpoint : " << endpoint << " Payload Size : " << data.size() << " HEX : " << data.toHex() << "STR : " << QString(data);
+
+        emit onDataReadFinished(endpoint, data);
 
-    onDataReceived(bt_socket);
+        onDataReceived(bt_socket);
+    }
 }
 
 void Pebble::onDataReadFinished(quint16 endPoint, const QByteArray &payload){
     //PhoneVersion is received immediatly after connecting to Pebble. We must respond to it first.
-    if(!pebbleReady && endPoint == Enums::Endpoint::PhoneVersion){
-        QByteArray payloadToSend;
-        QDataStream dataStream(&payloadToSend, QIODevice::WriteOnly);
-        dataStream << (quint8)PEBBLE_CLIENT_VERSION << (quint8)0xFF << (quint8)0xFF << (quint8)0xFF << (quint8)0xFF;
-        dataStream << (quint32) Enums::Session::GAMMA_RAY;
-        dataStream << (quint32) (Enums::Client::SMS | Enums::Client::TELEPHONY | Enums::Client::ANDROID);
-
-        sendDataToPebble(Enums::Endpoint::PhoneVersion, payloadToSend);
+    if(!pebbleReady && endPoint == Enums::Endpoint::PhoneVersion) {
+        unsigned int sessionCap = Enums::Session::GAMMA_RAY;
+        unsigned int remoteCap = Enums::Client::TELEPHONY | Enums::Client::SMS | Enums::Client::ANDROID;
+        QByteArray res;
+
+        //Prefix
+        res.append(0x01);
```

Comment:
Honestly i gave the ENUMS some thought and a try but i dont know what the values relate to, so i can define some magic values and use those enums instead but is that helping? do you have an idea how you would implement it? always up for suggestions